api/pkg/auth/google_auth.go

```diff
@@ -0,0 +1,76 @@
+package auth
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/idtoken"
+	"google.golang.org/api/option"
+	"google.golang.org/api/option/internaloption"
+	htransport "google.golang.org/api/transport/http"
+)
+
+// JSON key file types.
+const (
+	serviceAccountKey = "service_account"
+)
+
+type credentialsFile struct {
+	Type string `json:"type"`
+}
+
+// idTokenSource is an oauth2.TokenSource that wraps another TokenSource
+// It takes the id_token from TokenSource and passes that on as a bearer token
+type idTokenSource struct {
+	TokenSource oauth2.TokenSource
+}
+
+func (s *idTokenSource) Token() (*oauth2.Token, error) {
+	token, err := s.TokenSource.Token()
+	if err != nil {
+		return nil, err
+	}
+
+	idToken, ok := token.Extra("id_token").(string)
+	if !ok {
+		return nil, fmt.Errorf("token did not contain an id_token")
+	}
+
+	return &oauth2.Token{
+		AccessToken: idToken,
+		TokenType:   "Bearer",
+		Expiry:      token.Expiry,
+	}, nil
+}
+
+// InitGoogleClient is a helper method to be used by CaraML components to initialise a Google Client that appends ID
+// tokens to the headers of all outgoing requests with ID tokens, regardless of the type of credentials used
+func InitGoogleClient(ctx context.Context, audience string) (*http.Client, error) {
```

Comment:
Would the audience differ for the different MLP components? If not, should we also define the audience in this package for the components to refer to? Eg: `const audience="api.caraml"` ?

Reply:
> Would the audience differ for the different MLP components?

Hmm technically it shouldn't but it's an option if we want to have different audience values like `turing.api.caraml` or `merlin.api.caraml`, but yeah that might be an overkill, in which case specifying `api.caraml` as the audience uniformly throughout all the CaraML components would make sense here. I'll remove that audience argument then since it's highly unlikely we'll bother with working with so many different audience claims.